scene/gui/tree.cpp

```diff
@@ -247,31 +247,25 @@ void TreeItem::_propagate_check_through_parents(int p_column, bool p_emit_signal
 		return;
 	}
 
-	bool all_unchecked_and_not_indeterminate = true;
-	bool any_unchecked_or_indeterminate = false;
+	bool has_checked = false;
+	bool has_unchecked = false;
+	bool has_indeterminate = false;
 
 	TreeItem *child_item = current->get_first_child();
-	while (child_item) {
-		if (!child_item->is_checked(p_column)) {
-			any_unchecked_or_indeterminate = true;
-			if (child_item->is_indeterminate(p_column)) {
-				all_unchecked_and_not_indeterminate = false;
-				break;
-			}
+	while(child_item) {
+		if (child_item->is_checked(p_column)) {
+			has_checked = true;
+		} else if (child_item->is_indeterminate(p_column)) {
+			has_indeterminate = true;
 		} else {
-			all_unchecked_and_not_indeterminate = false;
+			has_unchecked = true;
 		}
 		child_item = child_item->get_next();
 	}
 
-	if (all_unchecked_and_not_indeterminate) {
-		current->set_checked(p_column, false);
-	} else if (any_unchecked_or_indeterminate) {
-		current->set_indeterminate(p_column, true);
-	} else {
-		current->set_checked(p_column, true);
-	}
-
+	current->set_checked(p_column, has_checked && !has_unchecked && !has_indeterminate);
+	current->set_indeterminate(p_column, has_indeterminate || (has_checked && has_unchecked));
```

Comment:
This logic should be done separately, as the two update each other, setting checked clears indeterminate and vice versa, they are exclusive

Reply:
Okay! will this work?

```
if (any_indeterminate || (any_checked && any_unchecked)){
	current->set_indeterminate(p_column, true);
} else {
	current->set_checked(p_column, any_checked);
}
```

Reply:
It should, as the second condition is only entered if `any_indeterminate` is false and `any_checked` and `any_unchecked` are not both true, so the condition should be equivalent.